cpp/src/arrow/compute/kernels/scalar_string.cc

```diff
@@ -427,6 +427,189 @@ void AddMatchSubstring(FunctionRegistry* registry) {
   DCHECK_OK(registry->AddFunction(std::move(func)));
 }
 
+// Slicing
+
+template <typename Type, typename Derived>
+struct SliceBase : StringTransform<Type, Derived> {
+  using Base = StringTransform<Type, Derived>;
+  using offset_type = typename Base::offset_type;
+  using State = OptionsWrapper<SliceOptions>;
+  SliceOptions options;
+
+  explicit SliceBase(SliceOptions options) : options(options) {}
+
+  static void Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    SliceOptions options = State::Get(ctx);
+    if (options.step == 0) {
+      ctx->SetStatus(Status::Invalid("Slice step cannot be zero"));
+      return;
+    }
+    Derived(options).Execute(ctx, batch, out);
+  }
+
+  void Execute(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    Base::Execute(ctx, batch, out);
+  }
+};
+
+#define PROPAGATE_FALSE(expr)         \
+  do {                                \
+    if (ARROW_PREDICT_FALSE(!expr)) { \
+      return false;                   \
+    }                                 \
+  } while (0)
+
+template <typename Type>
+struct SliceCodeunits : SliceBase<Type, SliceCodeunits<Type>> {
+  using Base = SliceBase<Type, SliceCodeunits<Type>>;
+  using offset_type = typename Base::offset_type;
+  using Base::Base;
+
+  bool Transform(const uint8_t* input, offset_type input_string_ncodeunits,
+                 uint8_t* output, offset_type* output_written) {
+    const uint8_t* begin = input;
+    const uint8_t* end = input + input_string_ncodeunits;
+    const uint8_t* begin_sliced = begin;
+    const uint8_t* end_sliced = end;
+    SliceOptions& options = this->options;
+
+    if (options.step >= 1) {
```

Comment:
The combinatory explosion in `if` branches seems a bit gratuitous? Abstractly it seems you should be able to do:
```c++
// Compute begin_sliced
if (options.start >= 0) {
  PROPAGATE_FALSE(
            arrow::util::UTF8AdvanceCodepoints(begin, end, &begin_sliced, options.start));
} else {
  PROPAGATE_FALSE(arrow::util::UTF8AdvanceCodepointsReverse(
            begin, end, &begin_sliced, -options.start));
}
// Compute end_sliced
if (options.stop >= 0) {
  PROPAGATE_FALSE(
            arrow::util::UTF8AdvanceCodepoints(begin, end, &end_sliced, options.stop));
} else {
  PROPAGATE_FALSE(arrow::util::UTF8AdvanceCodepointsReverse(
            begin, end, &end_sliced, -options.stop));
}
if (options.step > 0) {
  if (end_sliced < begin_sliced) {
    // empty slice
    *output_written = 0;
    return true;
  }
  // ...
}
else {
  DCHECK_LT(options.step, 0);
  if (end_sliced > begin_sliced) {
    // empty slice
    *output_written = 0;
    return true;
  }
  // ...
}
```

Reply:
The idea is that you don't start counting from the beginning or end twice when possible, e.g. if you do slice from -6 to -5, it's faster to not start counting code points from the end twice. It increases code complexity, that's the payoff.